src/test/librbd/test_librbd.cc

```diff
@@ -2574,6 +2574,238 @@ TEST_F(TestLibRBD, TestDataPoolIO)
   rados_ioctx_destroy(ioctx);
 }
 
+TEST_F(TestLibRBD, TestCompareAndWrite)
```

Comment:
Please split this into multiple tests, both by what is being tested and by sync vs async variant.  I see ten different tests here:

CompareAndWriteSuccess
AioCompareAndWriteSuccess
CompareAndWriteMismatch
AioCompareAndWriteMismatch
CompareAndWriteStripeUnitSuccess
AioCompareAndWriteStripeUnitSuccess
CompareAndWriteStripeUnitUnaligned
AioCompareAndWriteStripeUnitUnaligned
CompareAndWriteTooLarge
AioCompareAndWriteTooLarge

And similar for the C++ API, with `PP` suffix.

It will be somewhat repetitive, but that is fine.